### ISSUE TITLE
o_proj initialization

### ISSUE THREAD

**Issue by flozi00** [bjoernpl/bitllama#1]
Hi, nice to see some work in that area
I just checked the diff, to check why you are having the nans in training

In the llamaattention you are using this:

`self.o_proj = BitLinear(self.num_heads * self.head_dim, self.hidden_size, bias=config.attention_bias)`

while the transformers original is using:

`self.o_proj = nn.Linear(self.hidden_size, self.hidden_size, bias=config.attention_bias)`

Is there a specific reason for using num_heads * head_dim instead of hidden_size ?

**Comment by bjoernpl** [comment]
Good catch! Apparently I was working on an outdated fork of mine where I updated the main branch but forgot to update the feature branch. Not sure if this makes any difference regarding the problems I was getting but let me fix this.

**Comment by flozi00** [comment]
https://x.com/shxf0072/status/1771540151321104829?s=20

Just dropping it here, fyi
